File: backend/schemas/gtt.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


class GTTOut(BaseModel):
    trigger_id: int
    tradingsymbol: str
    trigger_type: str
    trigger_value: float | None
    limit_price: float | None
    quantity: int | None
    transaction_type: str
    status: str


class GTTsResponse(BaseModel):
    gtts: list[GTTOut]


class GTTCreateRequest(BaseModel):
    tradingsymbol: str
    exchange: Literal["NSE", "BSE", "NFO", "BFO", "MCX"] = "NSE"
    transaction_type: Literal["BUY", "SELL"]
    product: Literal["CNC", "MIS", "NRML"]
    trigger_type: Literal["single", "two-leg"]
    last_price: float
    # Single-leg
    trigger_value: float | None = None
    limit_price: float | None = None
    quantity: int | None = None
    # Two-leg
    upper_trigger_value: float | None = None
    upper_limit_price: float | None = None
    upper_quantity: int | None = None
    lower_trigger_value: float | None = None
    lower_limit_price: float | None = None
    lower_quantity: int | None = None


class GTTCreateResponse(BaseModel):
    trigger_id: int
    audit_log_id: str


class GTTModifyRequest(BaseModel):
    tradingsymbol: str
    exchange: Literal["NSE", "BSE", "NFO", "BFO", "MCX"] = "NSE"
    transaction_type: Literal["BUY", "SELL"]
    product: Literal["CNC", "MIS", "NRML"]
    last_price: float
    trigger_type: Literal["single", "two-leg"]
    trigger_value: float | None = None
    limit_price: float | None = None
    quantity: int | None = None
    upper_trigger_value: float | None = None
    upper_limit_price: float | None = None
    upper_quantity: int | None = None
    lower_trigger_value: float | None = None
    lower_limit_price: float | None = None
    lower_quantity: int | None = None


class GTTModifyResponse(BaseModel):
    trigger_id: int
    status: str

File: frontend/src/app/api/gtt.ts
import { apiFetch } from "./client";
import type {
  GTTListResponse,
  GTTCreateRequest,
  GTTModifyRequest,
  GTTPlaceResponse,
  GTTModifyResponse,
  ApiGTT,
} from "./types";
import type { GTTOrder } from "../data/mockData";

export function getGtts(): Promise<GTTListResponse> {
  return apiFetch<GTTListResponse>("/gtt");
}

export function placeGtt(body: GTTCreateRequest): Promise<GTTPlaceResponse> {
  return apiFetch<GTTPlaceResponse>("/gtt", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function modifyGtt(
  triggerId: number,
  body: GTTModifyRequest
): Promise<GTTModifyResponse> {
  return apiFetch<GTTModifyResponse>(`/gtt/${triggerId}`, {
    method: "PUT",
    body: JSON.stringify(body),
  });
}

export function deleteGtt(triggerId: number): Promise<void> {
  return apiFetch<void>(`/gtt/${triggerId}`, { method: "DELETE" });
}

// ---------------------------------------------------------------------------
// Mapper
// ---------------------------------------------------------------------------

export function mapGtt(g: ApiGTT): GTTOrder {
  return {
    id: String(g.trigger_id),
    symbol: g.tradingsymbol,
    type: g.trigger_type,
    transaction: g.transaction_type,
    quantity: g.quantity,
    status: g.status,
    // single-leg fields
    singleTrigger: g.trigger_value ?? undefined,
    singleLimit: g.limit_price ?? undefined,
    // two-leg fields
    upperTrigger: g.upper_trigger_value ?? undefined,
    upperLimit: g.upper_limit_price ?? undefined,
    lowerTrigger: g.lower_trigger_value ?? undefined,
    lowerLimit: g.lower_limit_price ?? undefined,
  };
}
